### PR TITLE
update scripts/ifAlias to work on FreeBSD as well as Linux

### PR DESCRIPTION
**Pull request by VVelox**
Allows interface descriptions set in FreeBSD to be read.

```
[root@vixen42]0|/usr/local/etc/namedb/master>ifconfig em3
em3: flags=8802<BROADCAST,SIMPLEX,MULTICAST> metric 0 mtu 1500
        description: this is a test
        options=81049b<RXCSUM,TXCSUM,VLAN_MTU,VLAN_HWTAGGING,VLAN_HWCSUM,LRO,VLAN_HWFILTER>
        ether 00:15:17:bf:9e:46
        media: Ethernet autoselect
        status: no carrier
        nd6 options=29<PERFORMNUD,IFDISABLED,AUTO_LINKLOCAL>
[root@vixen42]0|/usr/local/etc/namedb/master>
[kitsune@vixen42]0|/home/kitsune/github/librenms/scripts (fbsd-ifdesc $)>./ifAlias -n 4
.1.3.6.1.2.1.31.1.1.1.18.5
string
this is a test
```

Will pull the description config from there if set. :)

Also pointed it at "/etc/network/interfaces" if Ubuntu. Does not currently support netplan.

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [X] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [X] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [X] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by Jellyfrog** [comment]
Does ifconfig behave the same on bsd and Linux? 
In that case I would rather check if the binary exists instead of checking OS, then it works on other flavors of bsd also

**Comment by VVelox** [comment]
> Does ifconfig behave the same on bsd and Linux? In that case I would rather check if the binary exists instead of checking OS, then it works on other flavors of bsd also

Ifconfig on other unixes and Linux are two totally different thing. It is utterly unmaintained on Linux and increasingly not present in installs, having been replaced by ip where as it has under gone lots of development on other OSes.

On top of that using uname will make it easier to add support for other systems going forward.

That said checking the man files, description is supported on NetBSD and OpenBSD as well. Updating it now to only call uname once and also check if it is a BSD.

**Comment by VVelox** [comment]
Updated it to only call uname once and now operates in a generic manner on BSDs as looking at the docs it works the same on the other two as well. :)

File: scripts/ifAlias [hunk 1]
```diff
@@ -18,8 +18,17 @@ BASE='.1.3.6.1.2.1.31.1.1.1.18'
...
+COUNT="$(echo "${UNAME}" | grep -ci 'bsd$')"
+if [ "${COUNT}" -eq 1 ]; then
+    UNAME="BSD"
+fi
+
 # cache ip link output
-IP_LINK="$(ip link)"
+if [ "${UNAME}" = 'Linux' ]; then
+    IP_LINK="$(ip link)"
+fi
```
Comment (mpikzink) [inline scripts/ifAlias:31]: Do we have to cache this? We could thus do without a global variable.

File: scripts/ifAlias [hunk 2]
```diff
@@ -18,8 +18,17 @@ BASE='.1.3.6.1.2.1.31.1.1.1.18'
 GET_TYPE="$1"
 GET_OID="$2"
 
+UNAME="$(/usr/bin/uname)"
+
+COUNT="$(echo "${UNAME}" | grep -ci 'bsd$')"
+if [ "${COUNT}" -eq 1 ]; then
```
Comment (mpikzink) [inline scripts/ifAlias:24]: `if [ "$(echo "${UNAME}" | grep -ci 'bsd$')"  -eq 1 ] ; then`  We could thus do without a global `COUNT` variable. It's reassigned as local var in `alias_from_ip_link`

File: scripts/ifAlias [hunk 3]
```diff
@@ -51,20 +60,36 @@ interface_id()
...
-                    then
-                        echo -n "$N"
-                        return 0
-                    fi
-                    L="$N"
-                done
+                if [ "${UNAME}" = 'Linux' ]; then
+                    for N in $(echo "$IP_LINK" | grep -oE "^[0-9]+:" | cut -d':' -f 1)
+                    do
+                        if [ "$L" = "$ID" ] || [ ! "$ID" ]
```
Comment (mpikzink) [inline scripts/ifAlias:76]: ` if [ "$L" = "$ID" ] || [ -z  "$ID" ]`

File: scripts/ifAlias [hunk 4]
```diff
@@ -51,20 +60,36 @@ interface_id()
...
+                        then
+                            printf '%s' "$N"
+                            return 0
+                        fi
+                        L="$N"
+                    done
+                else
+                    for N in $(ifconfig | grep -E '^[A-Za-z]+' | cut -d: -f1 | cat -n -b | sed 's/^  *//' | sed 's/[\t\ ].*//'); do
+                        if [ "$L" = "$ID" ] || [ ! "$ID" ]
```
Comment (mpikzink) [inline scripts/ifAlias:85]: `if [ "$L" = "$ID" ] || [ -z "$ID" ]`

File: scripts/ifAlias [hunk 5]
```diff
@@ -73,42 +98,70 @@ interface_id()
...
-    local CONFIG_FILE=
+    CONFIG_FILE=
 
     if [ -x "$DISTRO_BIN" ]
     then
-        case "$("$DISTRO_BIN" | cut -d " " -f 1)" in
+        if [ "$(/usr/bin/uname)" = 'Linux' ]; then
+            DISTRO_VAR="$($DISTRO_BIN | cut -d " " -f 1)"
+        else
+            DISTRO_VAR="$(/usr/bin/uname)"
```
Comment (mpikzink) [inline scripts/ifAlias:117]: `DISTRO_VAR="$UNAME"`

File: scripts/ifAlias [hunk 6]
```diff
@@ -73,42 +98,70 @@ interface_id()
...
 alias_from_interfaces_config_file()
 {
-    local IFACE="$1"
-    local CONFIG_FILE=
+    CONFIG_FILE=
 
     if [ -x "$DISTRO_BIN" ]
     then
-        case "$("$DISTRO_BIN" | cut -d " " -f 1)" in
+        if [ "$(/usr/bin/uname)" = 'Linux' ]; then
```
Comment (mpikzink) [inline scripts/ifAlias:114]: `if [ "$UNAME" = 'Linux' ]; then"`

File: scripts/ifAlias [hunk 7]
```diff
@@ -73,42 +98,70 @@ interface_id()
...
-    local IFACE="$1"
-    ip link show "$IFACE" | grep -e "^[[:space:]]*alias[[:space:]]" | sed -e 's/^[[:space:]]*alias //'
+    case "${UNAME}" in
+        Linux)
+            ip link show "$1" | grep -e "^[[:space:]]*alias[[:space:]]" | sed -e 's/^[[:space:]]*alias //'
+            ;;
+        BSD)
+            COUNT=$(/sbin/ifconfig "$1" | grep 'description:' | head -n 1 | cut -d: -f 2- | wc -l)
+            if [ "${COUNT}" -eq 1 ]; then
+                /sbin/ifconfig "$1" | grep 'description:' | head -n 1 | cut -d: -f 2- | sed 's/^ //'
```
Comment (mpikzink) [inline scripts/ifAlias:158]:
```
LINK="$(/sbin/ifconfig "$1" | grep 'description:' | head -n 1 | cut -d: -f 2- | sed 's/^ //')"
if [ -n "$LINK" ]; then
echo "$LINK"
```

File: scripts/ifAlias [hunk 8]
```diff
@@ -73,42 +98,70 @@ interface_id()
...
+        else
+            DISTRO_VAR="$(/usr/bin/uname)"
+        fi
+
+        case "${DISTRO_VAR}" in
             Debian)
                 CONFIG_FILE="/etc/network/interfaces"
             ;;
+            Ubuntu)
+                CONFIG_FILE="/etc/network/interfaces"
```
Comment (mpikzink) [inline scripts/ifAlias:125]: We need to handle interfaces.d/* as well

**Review by mpikzink — COMMENTED** [review]
Nice work! I like it much. I haven't noticed about ifAliases in Linux. Have we a description in our doc how to set them with ip or in interfaces syntax?

File: scripts/ifAlias [hunk 9]
```diff
@@ -73,42 +98,70 @@ interface_id()
...
+                CONFIG_FILE="/etc/rc.conf"
             ;;
         esac
     fi
     if [ "$CONFIG_FILE" ]
     then
-         echo "$(grep -i "^# $IFACE:" $CONFIG_FILE | sed "s/^# $IFACE: //i")"
+         # echo squashes possible multi line replies to a single line
+         TO_ECHO_AND_MAKE_LING_HAPPY="$(grep -i "^# $1:" $CONFIG_FILE | sed "s/^# $1: //i")"
+         echo "${TO_ECHO_AND_MAKE_LING_HAPPY}"
```
[the remark this answers is not in the document]
Comment (VVelox) [reply scripts/ifAlias:145]: Depends on if we want to squash potential multi line returns down to a single line for config parsing. Went with the current operation which is to squash mutli line replies to a single line.

And aye. Docs should be updated. Grabbing that now.

File: scripts/ifAlias (hunk 1 above)
Comment (VVelox) [reply scripts/ifAlias:31]: Results in fewer external calls.

Really should do the same with ifconfig, but that can be done some other time.

**Comment by VVelox** [comment]
Okay. Should all be good now and every one happy. :)

**Comment by VVelox** [comment]
> Does ifconfig behave the same on bsd and Linux? In that case I would rather check if the binary exists instead of checking OS, then it works on other flavors of bsd also

To follow up on my earlier reply, just check Linux docs out of curiosity as it is has been forever since I've really used it on there(need ip for nearly everything I do with Linux) and yup, very much does not support the description field.

**Comment by murrant** [comment]
I have no idea what this script does, but yes, ifconfig should **not** be used on Linux.

File: scripts/ifAlias [hunk 10]
```diff
@@ -73,42 +97,77 @@ interface_id()
...
+    if [ "$CONFIG_FILE" ]; then
+         # echo squashes possible multi line replies to a single line
+         FOUND_LINES="$(grep -i "^# $1:" $CONFIG_FILE | sed "s/^# $1: //i")"
+         if [ "$(echo "${FOUND_LINES}" | wc -l)" -ge 1 ]; then
+             echo "${FOUND_LINES}"
+             return 0
+         fi
+    fi
+    if [ "${UNAME}" = "Linux" ] && [ -d '/etc/network/interfaces.d' ]; then
+         if [ "$(/bin/ls /etc/network/interfaces.d/ | wc -l)" -ge 1 ]; then
```
Comment (mpikzink) [inline scripts/ifAlias:148]: Better use find instead of ls

Comment (VVelox) [reply scripts/ifAlias]: Both work, pointless change made though. That said the next line did have an issue with grep.

**Comment by VVelox** [comment]
@mpikzink Anything else, I know I will likely regret asking as we have very different ideas on what needs changed? =^.^= :)

**Comment by murrant** [comment]
@VVelox I still can't figure out the purpose of scripts/ifAlias.  Care to explain what it is used for?

**Comment by VVelox** [comment]
@murrant Allows for interface descriptions akin to like on Cisco/Juniper devices. https://docs.librenms.org/Extensions/Interface-Description-Parsing/#setup

So this allows you to add something like 'Transit: ' etc for a interface and it will be picked up.

**Comment by Jellyfrog** [comment]
This script should probably move to the agent repo

**Comment by VVelox** [comment]
> This script should probably move to the agent repo

hmm, true! can make it so! :)

**Pull request closed without merging**